### PR TITLE
oop-principles / it-academy / JC1-117-22 / Ekaterina Komleva / Done all tasks

### PR DESCRIPTION
**Pull request by KateKoml**
## Pull request state

Actual status for PR:
- [x] implemented
- [x] reviewed
- [x] fixed
- [ ] approved

### CODE REVIEW

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Cat.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    public Cat(String name) {
+        this.name = name;
+    }
+
+    public String mew() {
+        return "mew-mew";
+    }
+
+    public void mew(Person man) {
+        double percentHappiness = -5;
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Cat.java:17]: move to class's constant

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Cat.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    public String purr() {
+        return "purrr-purrr-purrr";
+    }
+
+    public void purr(Person man) {
+        double percentHappiness = 5;
+        man.changeHappiness(percentHappiness);
+    }
+
+
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Cat.java:30]:
```suggestion

```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Cat.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Cat.java:39]:
```suggestion
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Persian.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package com.rakovets.course.java.core.practice.oop_principles.cat.cats;
+
+import com.rakovets.course.java.core.practice.oop_principles.cat.Person;
+import com.rakovets.course.java.core.practice.oop_principles.cat.cats.Cat;
+
+public class Persian extends Cat {
+
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Persian.java:7]:
```suggestion
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Persian.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    @Override
+    public String mew() {
+        return "Mya-mya-mya";
+    }
+
+    @Override
+    public void mew(Person man) {
+        double percentHappiness = -7;
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Persian.java:19]: etc

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Siamese.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package com.rakovets.course.java.core.practice.oop_principles.cat.cats;
+
+import com.rakovets.course.java.core.practice.oop_principles.cat.Person;
+import com.rakovets.course.java.core.practice.oop_principles.cat.cats.Cat;
+
+public class Siamese extends Cat {
+
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Siamese.java:7]:
```suggestion
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Sphynx.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package com.rakovets.course.java.core.practice.oop_principles.cat.cats;
+
+import com.rakovets.course.java.core.practice.oop_principles.cat.Person;
+import com.rakovets.course.java.core.practice.oop_principles.cat.cats.Cat;
+
+public class Sphynx extends Cat {
+
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Sphynx.java:7]:
```suggestion
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Sphynx.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    @Override
+    public String mew() {
+        return "Meww-meww-mewww";
+    }
+
+    @Override
+    public void mew(Person man) {
+        double percentHappiness = -25;
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/cat/cats/Sphynx.java:19]: etc

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class CatTests {
+
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java:19]:
```suggestion
```

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    @Test
+    void mew() {
+        assertEquals("mew-mew", new Cat("Bob").mew());
+        assertEquals("Miaau-miaau", new Siamese("Sim").mew());
+        assertEquals("Mya-mya-mya", new Persian("Pers").mew());
+        assertEquals("Meww-meww-mewww", new Sphynx("Nix").mew());
+    }
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java:26]: use Parametrized tests

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/heroes/Hero.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+    public void takeDamage(double damage) {
+        if (isAlive()) {
+            this.health -= damage;
+        }
+    }
+
+    public abstract String attackEnemy(Enemy enemy);
+
+    public abstract boolean isAlive();
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/heroes/Hero.java:31]: You need to add impl

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Enemy.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+    public void setHealth(double health) {
+        this.health = health;
+    }
+
+    public double getHealth() {
+        return health;
+    }
+
+
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Enemy.java:45]:
```suggestion
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Enemy.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    public void takeDamage(double damage) {
+        if (isAlive()) {
+            this.health -= damage;
+        }
+    }
+
+    public abstract String attackHero(Hero hero);
+
+    public abstract boolean isAlive();
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Enemy.java:31]: add impl

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/heroes/Mag.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
+package com.rakovets.course.java.core.practice.oop_principles.battle_ground.heroes;
+
+import com.rakovets.course.java.core.practice.oop_principles.battle_ground.enemies.Enemy;
+
+public class Mag extends Hero {
+    private double mana;
+    public Mag(String name) {
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/heroes/Mag.java:7]:
```suggestion
    private double mana;
    
    public Mag(String name) {
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/heroes/Warrior.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
+package com.rakovets.course.java.core.practice.oop_principles.battle_ground.heroes;
+
+import com.rakovets.course.java.core.practice.oop_principles.battle_ground.enemies.Enemy;
+
+public class Warrior extends Hero {
+    private double armor;
+    public Warrior(String name) {
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/heroes/Warrior.java:7]:
```suggestion
    private double armor;
    
    public Warrior(String name) {
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/heroes/Archer.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
+package com.rakovets.course.java.core.practice.oop_principles.battle_ground.heroes;
+
+import com.rakovets.course.java.core.practice.oop_principles.battle_ground.enemies.Enemy;
+
+public class Archer extends Hero {
+    private int fireArrowsNumber;
+    public Archer(String name) {
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/heroes/Archer.java:7]:
```suggestion
    private int fireArrowsNumber;
    
    public Archer(String name) {
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Demon.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
+package com.rakovets.course.java.core.practice.oop_principles.battle_ground.enemies;
+
+import com.rakovets.course.java.core.practice.oop_principles.battle_ground.heroes.Hero;
+
+public class Demon extends Enemy{
+    private int chanceToUseSuperAttack;
+    public Demon(double health) {
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Demon.java:7]:
```suggestion
    private int chanceToUseSuperAttack;
    
    public Demon(double health) {
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Werewolf.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+        double healingAbility = 5;
+        if (isAlive()) {
+            setHealth(getHealth() - damage);
+            if (getHealth() < lowHealthPoints) {
+                setHealth(getHealth() + healingAbility);
+            }
+        }
+    }
+
+    public boolean isAlive() {
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Werewolf.java:47]: use override

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Zombie.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
+package com.rakovets.course.java.core.practice.oop_principles.battle_ground.enemies;
+
+import com.rakovets.course.java.core.practice.oop_principles.battle_ground.heroes.Hero;
+
+public class Zombie extends Enemy{
+    private double maxHealth;
+    private int chanceToBeAlive;
+    public Zombie(double health) {
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Zombie.java:8]:
```suggestion
    private int chanceToBeAlive;
    
    public Zombie(double health) {
```

File: src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Zombie.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
+package com.rakovets.course.java.core.practice.oop_principles.battle_ground.enemies;
+
+import com.rakovets.course.java.core.practice.oop_principles.battle_ground.heroes.Hero;
+
+public class Zombie extends Enemy{
```
Comment (rakovets) [inline src/main/java/com/rakovets/course/java/core/practice/oop_principles/battle_ground/enemies/Zombie.java:5]:
```suggestion
public class Zombie extends Enemy {
```

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/BattleGroundTest.java [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+import com.rakovets.course.java.core.practice.oop_principles.battle_ground.heroes.Warrior;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+public class BattleGroundTest {
+    public static void main(String[] args) {
```
Comment (rakovets) [inline src/test/java/com/rakovets/course/java/core/practice/oop_principles/BattleGroundTest.java:19]: move to `Demo`

**Review by rakovets — CHANGES_REQUESTED** [review]
You need to fix

File: src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+    @MethodSource("provideArgumentsForPurr")
+    void testPurr(Cat cat, String expected) {
+        String actual = cat.purr();
+
+        Assertions.assertEquals(expected, actual);
+    }
+
+    static Stream<Arguments> provideArgumentsForMewByPerson() {
+        return Stream.of(
+                Arguments.of(new Person(50.0), new Cat("Bob"), 45.0),
```
Comment (KateKoml) [inline src/test/java/com/rakovets/course/java/core/practice/oop_principles/CatTests.java:54]:
```suggestion
                Arguments.of(new Person(50.0), new Cat("Bob"), 46.0),
```

**Pull request closed without merging**